### PR TITLE
scaffolding for 'implicitly available' packages in documents

### PR DESCRIPTION
**Pull request by kevinushey**
NOTE: Not quite ready for merge as this PR performs a few somewhat fundamental changes, but ready for discussion.

This PR refines the R source indexing logic in a couple of important ways.

Firstly, we now ensure that the R code is properly extracted from the associated document -- previously, the indexer just took e.g. RMarkdown content as-is, and so e.g.

    ```{r}
    some code
    ```

was actually being tokenized as three separate tokens: (``` `` ```, `` `{r}\nsome code\n` ``, ``` `` ```); ie, as three separate R symbols (!)

To alleviate this, the R indexer now only runs over the extracted R code, rather than the whole document.

In addition to this, some heuristics for detecting whether a particular package is 'implicitly' available for a document, e.g. for `runtime: shiny` in R Markdown documents, is now present as well.

### CODE REVIEW

**Comment by jjallaire** [comment]
Another thing to consider for Rmd is the new knit_params feature (whereby parameters and their defaults can be declared in yaml). As a result, references of the form `params$foo` should be considered valid. See https://github.com/yihui/knitr/pull/997 for details.

File: src/cpp/session/SessionRUtil.cpp [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+boost::regex& reYaml()
+{
+   static boost::regex instance("^[\\s\\n]*---\\s*(.*?)---\\s*(?:$|\\n)");
+   return instance;
+}
+
+} // anonymous namespace
+
+bool hasYamlHeader(const FilePath& filePath)
```
Comment (jjallaire) [inline src/cpp/session/SessionRUtil.cpp]: rmarkdown has a function to this which we could also call (that would keep us consistent, as it currently stands I think your regex is more liberal in what it considers front matter). The name of the function is partition_yaml_front_matter and it's here: https://github.com/rstudio/rmarkdown/blob/7cf3037663c8df05a82bd270cda52f5f702fd0af/R/output_format.R#L509

@jmcphers Also does some YAML handling on the server and perhaps one of these routines could be used.

Comment (kevinushey) [inline src/cpp/session/SessionRUtil.cpp]: Do we plan on keeping `rmarkdown` embedded (so that we can always be sure that this method is available)?

**Comment by kevinushey** [comment]
I think this is good to go now; since the only thing we do with the YAML header is scrape for `runtime: shiny` it's probably fine to leave it as-is with our own regex there.

**Pull request merged**